spec/ics-010-grandpa-client/README.md

```diff
@@ -0,0 +1,345 @@
+}
+```
+
+### Consensus state
+
+The GRANDPA client tracks authority set and commitment root for all previously verified consensus states.
+
+```typescript
+interface ConsensusState {
+  authoritySet: AuthoritySet
+  commitmentRoot: []byte
+}
+```
+
+### Headers
+
+The GRANDPA client headers include the height, the commitment root, a justification of block and a proof of authority set.
+
+```typescript
+interface Header {
+  height: uint64
+  commitmentRoot: []byte
+  justification: Justification
+  authoritySetProof: []byte
+}
+```
+
+### Justification
+
+A GRANDPA justification for block finality, it includes a commit message and an ancestry proof including all headers routing all precommit target blocks to the commit target block.
+
+```typescript
+interface Justification {
+  round: uint64
+  commit: Commit
+  votesAncestries: []Header
+}
+```
+
+### Evidence
+
+The `Evidence` type is used for detecting misbehaviour and freezing the client - to prevent further packet flow - if applicable.
+GRANDPA client `Evidence` consists of two headers at the same height both of which the light client would have considered valid.
+
+```typescript
+interface Evidence {
+  fromHeight: uint64
+  h1: Header
+  h2: Header
+}
+```
+
+### Client initialisation
+
+GRANDPA client initialisation requires a (subjectively chosen) latest consensus state, including the full authority set.
+
+```typescript
+function initialise(identifier: Identifier, height: uint64, consensusState: ConsensusState): ClientState {
+    set("clients/{identifier}/consensusStates/{height}", consensusState)
+    return ClientState{
+      latestHeight: height,
+      frozenHeight: null,
+    }
+}
+```
+
+The GRANDPA client `latestClientHeight` function returns the latest stored height, which is updated every time a new (more recent) header is validated.
+
+```typescript
+function latestClientHeight(clientState: ClientState): uint64 {
+  return clientState.latestHeight
+}
+```
+
+### Validity predicate
+
+GRANDPA client validity checking verifies a header is signed by the current authority set and verifies the authority set proof to determine if there is a expected change to the authority set. If the provided header is valid, the client state is updated & the newly verified commitment written to the store.
+
+```typescript
+function checkValidityAndUpdateState(
+  clientState: ClientState,
+  header: Header) {
+    // assert header height is newer than any we know
+    assert(header.height > clientState.latestHeight)
+    consensusState = get("clients/{identifier}/consensusStates/{clientState.latestHeight}")
+    // verify that the provided header is valid
+    assert(verify(header.justification, consensusState.authoritySet))
+    // update latest height
+    clientState.latestHeight = header.height
+    // verify that the authority set has been stored
+    assert(header.commitmentRoot.verifyMembership(path, authoritySet, header.authoritySetProof))
```

Comment:
`path` is undefined - also I don't understand the point of this, doesn't `verify` check that the authority set signed the header?

Reply:
Yes, it does.
Line 144 is used to get the latest authority set, I removed this line and added an explanation in the section of Headers.